README.md
[![fly.js](https://github.com/wendux/fly/raw/master/fly.png)](https://wendux.github.io/dist/#/doc/flyio/readme)

[![npm version](https://img.shields.io/npm/v/flyio.svg)](https://www.npmjs.org/package/flyio)
[![build status](https://travis-ci.org/wendux/fly.svg?branch=master)](https://travis-ci.org/wendux/fly)
[![coverage](https://img.shields.io/codecov/c/github/wendux/fly/master.svg)](https://codecov.io/github/wendux/fly?branch=master)
[![size](https://img.shields.io/github/size/wendux/fly/dist/fly.min.js.svg)](https://unpkg.com/flyio@0.3.1/dist/fly.min.js)
![platform](https://img.shields.io/badge/platforms-All%20JavaScript%20Runtimes-blue.svg)

## Fly.js

Supporting request forwarding and Promise based HTTP client for all JavaScript runtimes.

Chinese documentation : [中文文档](https://github.com/wendux/fly/blob/master/README-CH.md)



### Browser Support

| ![Chrome](https://raw.github.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png) | ![Firefox](https://raw.github.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png) | ![Safari](https://raw.github.com/alrra/browser-logos/master/src/safari/safari_48x48.png) | ![Opera](https://raw.github.com/alrra/browser-logos/master/src/opera/opera_48x48.png) | ![Edge](https://raw.github.com/alrra/browser-logos/master/src/edge/edge_48x48.png) | ![IE](https://raw.github.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png) |
| ---------------------------------------- | ---------------------------------------- | ---------------------------------------- | ---------------------------------------- | ---------------------------------------- | ---------------------------------------- |
| ✔                                        | ✔                                        | ✔                                        | ✔                                        | ✔                                        | > 8                                      |

### Other Platforms  Support
<table>
    <tbody>
    <tr>
        <td align="center" valign="middle">
            <a href="https://nodejs.org/">
                <img  src="https://github.com/wendux/fly/raw/master/imgs/node.png" alt="node logo">
            </a>
        </td>
        <td align="center" valign="middle">
            <a href="https://mp.weixin.qq.com/">
                <img  src="https://github.com/wendux/fly/raw/master/imgs/mp.png" alt="Mini Program logo">
            </a>
        </td>
        <td align="center" valign="middle">
            <a href="http://facebook.github.io/react-native/">
                <img  src="https://github.com/wendux/fly/raw/master/imgs/rn.png" alt="mpvue logo">
            </a>
        </td>
        <td align="center" valign="middle">
            <a href="http://weex.apache.org/">
                <img  src="https://github.com/wendux/fly/raw/master/imgs/weex.png" alt="weex logo">
            </a>
        </td>
        <td align="center" valign="middle">
            <a href="https://www.quickapp.cn/">
                <img  src="https://github.com/wendux/fly/raw/master/imgs/hp.png" alt="quick app logo">
            </a>
        </td>
    </tr>
    </tbody>
</table>  

Currently the platforms that fly.js supported include  [Node.js](https://nodejs.org/) 、[WeChat Mini Program](https://mp.weixin.qq.com/cgi-bin/wx) 、[Weex](http://weex.apache.org/) 、[React Native](http://facebook.github.io/react-native/) 、[Quick App](https://www.quickapp.cn/) and the browers, all  JavaScript runtimes of these are different. More platforms  are supporting...

## Features

1. Supports the [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) API
2. Make [XMLHttpRequests](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) from the browser，**light-weight and very light-weight**
3. Supports  various  JavaScript runtimes
4. Supports  request and response interceptors。
5. Automatic transforms for JSON data。
6. Supports switching the underlying Http Engine, easy to adapt to various JavaScript Runtime.
7. Supports global Ajax interception on browser
8. Supports request forwarding in hybrid applications




## Positioning & target

Fly.is  locates to be the ultimate solution for Javascript http requests. That is to say, in any environment that can execute Javascript, as long as it has the ability to access the network, Fly can run on it and provide unified APIs. At the same time, keep lightweight on the browser side.



## Documentation

You can find the Fly documentation [on the offical website](https://wendux.github.io/dist/#/language).

[中文文档](https://wendux.github.io/dist/#/doc/flyio/readme)

## Installing

Using npm 

```shell
npm install flyio
```

Using CDN （on browsers）

```javascript
<script src="https://unpkg.com/flyio/dist/fly.min.js"></script>
```

UMD （on browsers）

```http
https://unpkg.com/flyio/dist/umd/fly.umd.min.js
```



## Require flyio

The entry files of  different JavaScript Runtimes may be different, you can refer on the below，
but the entries of  Browsers, Node, and React Native are the same。each requirement  ways  of these platforms are as follows:

1. Requiring On browsers、Node、React Native

   ```javascript
   //for Browsers, Node, and React Native
   var fly=require("flyio")
   ```

   The above method requires the default instance of Fly, and you can also create an Fly instance by yourself:

   ```javascript
   // for browsers and React Native
   var Fly=require("flyio/dist/npm/fly")
   // for Node
   //var Fly=require("flyio/src/node")
   var fly=new Fly;
   ```

2. Requiring on WeChat Mini Program

   ```javascript
   var Fly=require("flyio/dist/npm/wx") 
   var fly=new Fly
   ```

   If your  project does not use NPM to manage dependencies, you can download the source code directly to your project 。The download links are [wx.js](https://github.com/wendux/fly/tree/master/dist/npm/wx.js) or  [wx.umd.min.js](https://github.com/wendux/fly/tree/master/dist/umd/wx.umd.min.js)  . Download any one, save it to your project directory (assuming named "lib") , and then requiring:

   ```javascript
   var Fly=require("../lib/wx") 
   var fly=new Fly; 
   ```

3. Requiring on Quick App

   On [Quick App](https://www.quickapp.cn/), Fly relies on the `fetch` module, which needs to be defined in the manifest.json first:

   ```java
     "features": [
        ...
       {"name": "system.fetch"}
     ]
   ```

   And then create Fly instance:

   ```javascript
    var fetch = require("@system.fetch")
    var Fly=require("flyio/dist/npm/hap")
    var fly=new Fly(fetch)
   ```

4. Requiring on Weex

   ```javascript
   var Fly=require("flyio/dist/npm/weex")
   var fly=new Fly
   ```

After the requirement, you can make the global configuration and add the interceptor to `fly` , and peform network request with `fly`.

## Example

The following example, if not specified, can be executed in all JavaScript Runtimes.

### Performing a `GET` request

```javascript
var fly=require("flyio")
// Make a request for a user with a given ID,  and the parameter is passed directly in URL
fly.get('/user?id=133')
  .then(function (response) {
    console.log(response);
  })
  .catch(function (error) {
    console.log(error);
  });

// The parameter is passed by a object
fly.get('/user', {
      id: 133
  })
  .then(function (response) {
    console.log(response);
  })
  .catch(function (error) {
    console.log(error);
  });
```

### Performing a `POST` request

```javascript
fly.post('/user', {
    name: 'Doris',
    age: 24
    phone:"18513222525"
  })
  .then(function (response) {
    console.log(response);
  })
  .catch(function (error) {
    console.log(error);
  });
```

### Performing multiple concurrent requests

```javascript
function getUserRecords() {
  return fly.get('/user/133/records');
}

function getUserProjects() {
  return fly.get('/user/133/projects');
}

fly.all([getUserRecords(), getUserProjects()])
  .then(fly.spread(function (records, projects) {
    // Both requests are now complete
  }))
  .catch(function(error){
    console.log(error)
  })
```

### Performing the request by `request`

```javascript
fly.request("/test",{hh:5},{
    method:"post",
    timeout:5000 // Set timeout to 5 seconds
 })
.then(d=>{ console.log("request result:",d)})
.catch((e) => console.log("error", e))
```



### Sending `URLSearchParams`

```javascript
const params = new URLSearchParams();
params.append('a', 1);
fly.post("",params)
.then(d=>{ console.log("request result:",d)})
```

> Note that `URLSearchParams` is not supported by all browsers (see [caniuse.com](http://www.caniuse.com/#feat=urlsearchparams)), but there is a [polyfill](https://github.com/WebReflection/url-search-params) available (make sure to polyfill the global environment).

### Sending `FormData`

```javascript
 var formData = new FormData();
 var log=console.log
 formData.append('username', 'Chris');
 fly.post("../package.json",formData).then(log).catch(log)
```

> Note that `FormData` is not supported by all browsers (see [caniuse.com](http://www.caniuse.com/#feat=urlsearchparams)), but there is a [polyfill](https://github.com/form-data/form-data) available (make sure to polyfill the global environment).  And there are some differences on the way to support formData in node environment. Please click [here](#/doc/flyio/node) for more details

### Requesting Stream

```javascript
fly.get("/Fly/v.png",null,{
	responseType:"arraybuffer"
}).then(d=>{
  //d.data is a ArrayBuffer instance
})
```

> In browser, the value of responseType can be one of "arraybuffer" or "blob". In node, you can just set it as "stream"
>

## interceptors

You can intercept requests or responses before they are handled by `then` or `catch`.

```javascript

// Add a request interceptor
fly.interceptors.request.use((request)=>{
    // Do something before request is sent
    request.headers["X-Tag"]="flyio";
  	console.log(request.body)
    // Complete the request with custom data
    // return Promise.resolve("fake data")
})

// Add a response interceptor
fly.interceptors.response.use(
    (response) => {
      // Do something with response data .
      // Just return the data field of response
      return response.data
    },
    (err) => {
      // Do something with response error
      //return Promise.resolve("ssss")
    }
)
```

The structures of the `request ` object in request interceptor.

```javascript
{
  baseURL,  //base url
  body, // request parameters
  headers, //custom request headers
  method, // http request method
  timeout, // request time
  url, // request url (or relative path)
  withCredentials, // determine whether sending thirdparty cookies in cross-domain request
  ... // custom field defined in options
}
```

The structures of the `response` object in response interceptor.

```javascript
{
  data, //response data
  engine, //http engine,In browser,it's a instance of XMLHttpRequest.
  headers, //response headers
  request  //the origin request object
}
```

### Remove interceptors 

If you may need to remove an interceptor later,    just set it to null.

```javascript
fly.interceptors.request.use(null)
fly.interceptors.response.use(null,null)
```

### Perform  an async task in  interceptors

Now,  you can perform  async task in interceptors !

Let's see an example:

because of security reasons, we need all the requests to set up a csrfToken in the header, if csrfToken does not exist, we need to request a csrfToken first, and then perform the network request, because the request csrfToken progress is asynchronous, so we need to execute  this async request in request interceptor. the code is as follows:

```javascript
var csrfToken="";
var tokenFly=new Fly();
var fly=new Fly();
fly.interceptors.request.use(function (request) {
    //if csrfToken does not exist, we need to request a csrfToken first
    if(!csrfToken) {
         // locking the current instance, let the incomming request task enter a 
         // queue before they enter the request interceptors.
         fly.lock();
         //Using  another fly instance to request csrfToken.
         //If use the same fly instance, there may lead a infinite loop:
         //(The request will go to the interceptor first, and then
         //enter the interceptor again when launching the new request 
         //in the interceptor....)
         return tokenFly.get("/token").then((d)=>{
            request.headers["csrfToken"]=csrfToken=d.data.data.token;
            //only return the origin `request` object can make the http request continue.
            // otherwise, the return data will be teated as "response" data.
            return request 
          }).finally(()=>{
           //fly.clear(); //clear the request queue
           // unlock the current instance, flush the request queue.   
           fly.unlock()
         })  
        
    }else {
        request.headers["csrfToken"]= csrfToken;
        //This line can be omitted. 
        //If the interceptor doesn't return value, `request` will be used by default.
        return request 
    }
})
```

**Note**: 

1. The current fly instance will be locked  when call `fly.lock()` . Once  the fly instance is locked,  the incomming request task maked by it will be hang up and enter a queue before they enter the request interceptors, you can call `fly.unlock()` to continue the requests or call `fly.clear()` to cancel the requests in the queue.
2. **Only when you return the `request` object  passed by interceptor at the final , the origin http request  will be continued**.  

And you can also make an async task in the **response** interceptor. More information about interceptors and examples refer to [flyio interceptor](https://wendux.github.io/dist/#/doc/flyio/interceptor).

## Error handling

If  the request fails, `catch`  will be called;  the error object's fields as follows :

```javascript
{
  message:"Not Find 404", //error description
  status:404, // error code
  request:{...}, //the request info
  response:{...}, //the error response,
  engine:{}, // The Http engine, it's a instance of XmlHttpRequest in browser.
}
```

| Code  | Description                              |
| ----- | ---------------------------------------- |
| 0     | network error                            |
| 1     | timeout                                  |
| 2     | The file was downloaded successfully, but the save failed. **This error only occurred in the node environment** |
| >=200 | http status code                         |

**Example**

```javascript
fly.get('/user/12345')
  .catch(function (error) {
    console.log('Error', error.message);
    console.log('Error engine:', error.engine)
    // The request was made but no response was received
    // `error.request` holds the request info.
    console.log('Error request info:',error.request);
    if (error.response) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx and 304
      console.log(error.response.data);
      console.log(error.response.status);
      console.log(error.response.statusText);
      console.log(error.response.headers);
    }
  });
```

## Request configuration options

**Configurable options**：

```javascript
{
 // `method` is the request method to be used when making the request
  method: 'get', // default
  // `headers` are custom headers to be sent
  headers:{},
  // `baseURL` will be prepended to `url` unless `url` is absolute.
  // It can be convenient to set `baseURL` for an instance of fly to pass relative URLs
  // to methods of that instance.
  baseURL:"",
  // `timeout` specifies the number of milliseconds before the request times out.
  // If the request takes longer than `timeout`, the request will be aborted.
  timeout:0,  //default
  // `parseJson` indicates whether or not it is  automatically converted response 
  //  data , the Content-Type of which is application/json, to JSON object,  
  parseJson:true,
  // `withCredentials` indicates whether or not cross-site Access-Control requests
  // should be made using credentials
  withCredentials: false, // default
}
```

Configuration supports **instance level configuration** and **single request configuration**

### Instance level

Instance level configuration can be applied to all requests initiated by the current Fly instance:

```javascript
// Add common headers
fly.config.headers={xx:5,bb:6,dd:7}
// Set timeout
fly.config.timeout=10000;
// Set base url
fly.config.baseURL="https://wendux.github.io/"
```

### Single request configuration

Single request configuration is valid only once:

```javascript
fly.request("/test",{hh:5},{
    method:"post",
    timeout:5000 //超时设置为5s
})
```

> If the single configuration and the instance configuration conflict, the single request configuration is preferred .

## API

#### `fly.get(url, data, options)`

#### `fly.post(url, data, options)`

In browser environment, the type of data can be:

```shell
String|Json|Object|Array|Blob|ArrayBuffer|FormData
```

`options` is the  request configuration。



#### `fly.request(url, data, options)`

When you use this API, you should explicitly specify  the method of options, and the default method is `get`

```javascript
// Perform a get request
fly.request("/user/8" null, {method:"get"})
// Perform a delete request
fly.request("/user/8/delete", null, {method:"delete"})
// Perform a put request
fly.request("/user/register", {name:"doris"}, {method:"PUT"})
......
```

This API is suitable for use in RESTful API scenarios. For convenience, the corresponding alias methods are provided
**alias**

`fly.put(url, data, options)`

`fly.delete(url,data,options)`

`fly.patch(url,data,options)`

### Concurrency

Helper functions for dealing with concurrent requests.

##### `fly.all([])`
##### `fly.spread(callback)`

The parameter of all is an array of `promise`,  `then` will be called only when all requests are successful, and` catch` will be called as long as one fails.

## Using application/x-www-form-urlencoded format

By default, fly serializes JavaScript objects to `JSON`. To send data in the `application/x-www-form-urlencoded` format instead, you can use one of the following options.

### By specifying `content-type` request header

When you specify  `content-type` value as "application/x-www-form-urlencoded", fly will encode the request data automatically, for example：

```javascript
fly.post("../package.json",{aa:8,bb:9,tt:{xx:5}},{headers:{
    "content-type":"application/x-www-form-urlencoded"
}})
.then(console.log)
```

This method is universal, which means that it can work well in any JavaScript runtime. There are also some other ways for some specific platforms:

### Other ways

#### Browser

In a browser, you can use the [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) API as follows:

```js
var params = new URLSearchParams();
params.append('param1', 'value1');
params.append('param2', 'value2');
fly.post('/foo', params);
```

> Note that `URLSearchParams` is not supported by all browsers (see [caniuse.com](http://www.caniuse.com/#feat=urlsearchparams)), but there is a [polyfill](https://github.com/WebReflection/url-search-params) available (make sure to polyfill the global environment).

Alternatively, you can encode data using the [`qs`](https://github.com/ljharb/qs) library:

```js
var qs = require('qs');
fly.post('/foo', qs.stringify({ 'bar': 123 }));
```

#### Node.js

In node.js, you can use the [`querystring`](https://nodejs.org/api/querystring.html) module as follows:

```js
var querystring = require('querystring');
fly.post('http://something.com/', querystring.stringify({ foo: 'bar' }));
```

You can also use the [`qs`](https://github.com/ljharb/qs) library.



## Promises

Fly depends on a native ES6 Promise implementation to be [supported](http://caniuse.com/promises).
If your environment doesn't support ES6 Promises, you can [polyfill](https://github.com/jakearchibald/es6-promise).

## TypeScript
fly includes [TypeScript](http://typescriptlang.org) definitions.
```typescript
import fly from "flyio"
fly.get('/user?ID=12345');
```
## Creating  an instance

You can create a new instance of Fly , and then apply different configurations:

```javascript
// Notice require path is  "flyio/dist/npm/fly"
var  Fly=require("flyio/dist/npm/fly")
var nFly=new Fly();
// Apply different configurations
nFly.config.timeout=15000
nfly.interceptors.request.use(...)
```



## Http Engine


CHANGELOG.md
# Changelog

### **v0.6.1** (Sep 6, 2018)

+ 添加“params”配置项：
 - 可以设置公共的的get参数：fly.config.params={token:"xxxx"}
 - 非Get请求也可以设置Url params
 - Get请求的话会合并data和“params”

- 在非浏览器环境下，响应头中有多个"set-cookie"字段时，headers["set-cookie"]值为数组类型(之前为字符串)

### **v0.5.4** (Apr 20, 2018)

+ Adding `lock/unlock` API for interceptors and fly propotype.
+ update typings file for typescript
+ delete ~~`await`~~ that added in v0.5.2.



src/fly.js
var utils = require('./utils/utils');
var isBrowser = typeof document !== "undefined";

class Fly {
    constructor(engine) {
        this.engine = engine || XMLHttpRequest;

        this.default = this   //For typeScript

        /**
         * Add  lock/unlock API for interceptor.
         *
         * Once an request/response interceptor is locked, the incoming request/response
         * will be added to a queue before they enter the interceptor, they will not be
         * continued  until the interceptor is unlocked.
         *
         * @param [interceptor] either is interceptors.request or interceptors.response
         */
        function wrap(interceptor) {
            var resolve;
            var reject;

            function _clear() {
                interceptor.p = resolve = reject = null;
            }

            utils.merge(interceptor, {
                lock() {
                    if (!resolve) {
                        interceptor.p = new Promise((_resolve, _reject) => {
                            resolve = _resolve
                            reject = _reject;
                        })
                    }
                },
                unlock() {
                    if (resolve) {
                        resolve()
                        _clear();
                    }
                },
                clear() {
                    if (reject) {
                        reject("cancel");
                        _clear();
                    }
                }
            })
        }

        var interceptors = this.interceptors = {
            response: {
                use(handler, onerror) {
                    this.handler = handler;
                    this.onerror = onerror;
                }
            },
            request: {
                use(handler) {
                    this.handler = handler;
                }
            }
        }

        var irq = interceptors.request;
        var irp = interceptors.response;
        wrap(irp);
        wrap(irq);

        this.config = {
            method: "GET",
            baseURL: "",
            headers: {},
            timeout: 0,
            params: {}, // Default Url params
            parseJson: true, // Convert response data to JSON object automatically.
            withCredentials: false
        }
    }

    request(url, data, options) {
        var engine = new this.engine;
        var contentType = "Content-Type";
        var contentTypeLowerCase = contentType.toLowerCase();
        var interceptors = this.interceptors;
        var requestInterceptor = interceptors.request;
        var responseInterceptor = interceptors.response;
        var requestInterceptorHandler = requestInterceptor.handler;
        var promise = new Promise((resolve, reject) => {
            if (utils.isObject(url)) {
                options = url;
                url = options.url;
            }
            options = options || {};
            options.headers = options.headers || {};

            function isPromise(p) {
                // some  polyfill implementation of Promise may be not standard,
                // so, we test by duck-typing
                return p && p.then && p.catch
            }

            /**
             * If the request/response interceptor has been locked，
             * the new request/response will enter a queue. otherwise, it will be performed directly.
             * @param [promise] if the promise exist, means the interceptor is  locked.
             * @param [callback]
             */
            function enqueueIfLocked(promise, callback) {
                if (promise) {
                    promise.then(() => {
                        callback()
                    })
                } else {
                    callback()
                }
            }

            // make the http request
            function makeRequest(options) {
                data = options.body;
                // Normalize the request url
                url = utils.trim(options.url);
                var baseUrl = utils.trim(options.baseURL || "");
                if (!url && isBrowser && !baseUrl) url = location.href;
                if (url.indexOf("http") !== 0) {
                    var isAbsolute = url[0] === "/";
                    if (!baseUrl && isBrowser) {
                        var arr = location.pathname.split("/");
                        arr.pop();
                        baseUrl = location.protocol + "//" + location.host + (isAbsolute ? "" : arr.join("/"))
                    }
                    if (baseUrl[baseUrl.length - 1] !== "/") {
                        baseUrl += "/"
                    }
                    url = baseUrl + (isAbsolute ? url.substr(1) : url)
                    if (isBrowser) {

                        // Normalize the url which contains the ".." or ".", such as
                        // "http://xx.com/aa/bb/../../xx" to "http://xx.com/xx" .
                        var t = document.createElement("a");
                        t.href = url;
                        url = t.href;
                    }
                }

                var responseType = utils.trim(options.responseType || "")
                var isGet = options.method === "GET";
                var dataType = utils.type(data);
                var params = options.params || {};

                // merge url params when the method is "GET" (data is object)
                if (isGet && dataType === "object") {
                    params = utils.merge(data, params)
                }
                // encode params to String
                params = utils.formatParams(params);

                // save url params
                var _params = [];
                if (params) {
                    _params.push(params);
                }
                // Add data to url params when the method is "GET" (data is String)
                if (isGet && data && dataType === "string") {
                    _params.push(data);
                }

                // make the final url
                if (_params.length > 0) {
                    url += (url.indexOf("?") === -1 ? "?" : "&") + _params.join("&");
                }

                engine.open(options.method, url);

                // try catch for ie >=9
                try {
                    engine.withCredentials = !!options.withCredentials;
                    engine.timeout = options.timeout || 0;
                    if (responseType !== "stream") {
                        engine.responseType = responseType
                    }
                } catch (e) {
                }

                var customContentType = options.headers[contentType] || options.headers[contentTypeLowerCase];

                // default content type
                var _contentType = "application/x-www-form-urlencoded";
                // If the request data is json object, transforming it  to json string,
                // and set request content-type to "json". In browser,  the data will
                // be sent as RequestBody instead of FormData
                if (utils.trim((customContentType || "").toLowerCase()) === _contentType) {
                    data = utils.formatParams(data);
                } else if (!utils.isFormData(data) && ["object", "array"].indexOf(utils.type(data)) !== -1) {
                    _contentType = 'application/json;charset=utf-8'
                    data = JSON.stringify(data);
                }
                //If user doesn't set content-type, set default.
                if (!customContentType) {
                    options.headers[contentType] = _contentType;
                }

                for (var k in options.headers) {
                    if (k === contentType && utils.isFormData(data)) {
                        // Delete the content-type, Let the browser set it
                        delete options.headers[k];
                    } else {
                        try {
                            // In browser environment, some header fields are readonly,
                            // write will cause the exception .
                            engine.setRequestHeader(k, options.headers[k])
                        } catch (e) {
                        }
                    }
                }

                function onresult(handler, data, type) {
                    enqueueIfLocked(responseInterceptor.p, function () {
                        if (handler) {
                            //如果失败，添加请求信息
                            if (type) {
                                data.request = options;
                            }
                            var ret = handler.call(responseInterceptor, data, Promise)
                            data = ret === undefined ? data : ret;
                        }
                        if (!isPromise(data)) {
                            data = Promise[type === 0 ? "resolve" : "reject"](data)
                        }
                        data.then(d => {
                            resolve(d)
                        }).catch((e) => {
                            reject(e)
                        })
                    })
                }


                function onerror(e) {
                    e.engine = engine;
                    onresult(responseInterceptor.onerror, e, -1)
                }

                function Err(msg, status) {
                    this.message = msg
                    this.status = status;
                }

                engine.onload = () => {
                    // The xhr of IE9 has not response field
                    var response = engine.response || engine.responseText;
                    if (response && options.parseJson && (engine.getResponseHeader(contentType) || "").indexOf("json") !== -1
                        // Some third engine implementation may transform the response text to json object automatically,
                        // so we should test the type of response before transforming it
                        && !utils.isObject(response)) {
                        response = JSON.parse(response);
                    }

                    var headers = engine.responseHeaders;
                    // In browser
                    if (!headers) {
                        headers = {};
                        var items = (engine.getAllResponseHeaders() || "").split("\r\n");
                        items.pop();
                        items.forEach((e) => {
                            if (!e) return;
                            var key = e.split(":")[0]
                            headers[key] = engine.getResponseHeader(key)
                        })
                    }
                    var status = engine.status
                    var statusText = engine.statusText
                    var data = {data: response, headers, status, statusText};
                    // The _response filed of engine is set in  adapter which be called in engine-wrapper.js
                    utils.merge(data, engine._response)
                    if ((status >= 200 && status < 300) || status === 304) {
                        data.engine = engine;
                        data.request = options;
                        onresult(responseInterceptor.handler, data, 0)
                    } else {
                        var e = new Err(statusText, status);
                        e.response = data
                        onerror(e)
                    }
                }

                engine.onerror = (e) => {
                    onerror(new Err(e.msg || "Network Error", 0))
                }

                engine.ontimeout = () => {
                    onerror(new Err(`timeout [ ${engine.timeout}ms ]`, 1))
                }
                engine._options = options;
                setTimeout(() => {
                    engine.send(isGet ? null : data)
                }, 0)
            }

            enqueueIfLocked(requestInterceptor.p, () => {
                utils.merge(options, this.config)
                var headers = options.headers;
                headers[contentType] = headers[contentType] || headers[contentTypeLowerCase] || "";
                delete headers[contentTypeLowerCase]
                options.body = data || options.body;
                url = utils.trim(url || "");
                options.method = options.method.toUpperCase();
                options.url = url;
                var ret = options;
                if (requestInterceptorHandler) {
                    ret = requestInterceptorHandler.call(requestInterceptor, options, Promise) || options;
                }
                if (!isPromise(ret)) {
                    ret = Promise.resolve(ret)
                }
                ret.then((d) => {
                    //if options continue
                    if (d === options) {
                        makeRequest(d)
                    } else {
                        resolve(d)
                    }
                }, (err) => {
                    reject(err)
                })
            })
        })
        promise.engine = engine;
        return promise;
    }

    all(promises) {
        return Promise.all(promises)
    }

    spread(callback) {
        return function (arr) {
            return callback.apply(null, arr);
        }
    }
}

//For typeScript
Fly.default = Fly;

["get", "post", "put", "patch", "head", "delete"].forEach(e => {
    Fly.prototype[e] = function (url, data, option) {
        return this.request(url, data, utils.merge({method: e}, option))
    }
});
["lock", "unlock", "clear"].forEach(e => {
    Fly.prototype[e] = function () {
        this.interceptors.request[e]();
    }
});
// Learn more about keep-loader: https://github.com/wendux/keep-loader
KEEP("cdn||cdn-min", () => {
    // This code block will be removed besides the  "CDN" and "cdn-min" build environment
    window.fly = new Fly;
    window.Fly = Fly;
})
module.exports = Fly;


